ee/clickhouse/sql/trends/lifecycle.py

```diff
@@ -1,184 +1,129 @@
+            ASOF LEFT JOIN unbounded_filtered_events previous_activity
+                ON previous_activity.person_id = target.person_id 
+                    AND target.start_of_period > previous_activity.timestamp
 
-    GROUP BY 
-        person_id, 
-        initial_period
-
-    UNION ALL
-
-    /* 
-        Get the first active period for each user within the extended range 
-        i.e. including the previous period
-        
-        NOTE: initial_period and next_period are the same
-    */ 
-    SELECT
-        base.person_id, 
-        min(base.start_of_period) as initial_period, 
-        min(base.start_of_period) as next_period 
-        
-    FROM person_activity_including_previous_period base
-    GROUP BY person_id
-
-    UNION ALL
-
-    /*
-        Get activity status rows for all dormant periods and all persons
-    */
-    SELECT 
-        person_id, 
-        initial_period, 
-        next_period 
+        UNION ALL
 
-    FROM (
+        /*
+            Get periods just after activity, and classify them as 'dormant'
+        */
         SELECT 
-            person_activity.person_id AS person_id, 
-
-            -- Use datetime null value as marker that this refers to dormant
-            toDateTime('0000-00-00 00:00:00') as initial_period, 
-            periods.start_of_period as next_period
-
-        FROM person_activity_as_array as person_activity
-            CROSS JOIN periods
-
-        WHERE has(person_activity.start_of_period, periods.start_of_period) = 0
+            person_id,
+            activity_before_target.start_of_period + interval_type AS start_of_period,
+            'dormant' AS status
+
+        FROM bounded_person_activity_by_period activity_before_target
+            ASOF LEFT JOIN unbounded_filtered_events next_activity
+                ON activity_before_target.person_id = next_activity.person_id 
+                    AND next_activity.timestamp > activity_before_target.start_of_period + interval_type
+
+            WHERE next_activity.person_id = '00000000-0000-0000-0000-000000000000'
+                OR dateDiff(
+                    selected_period, 
+                    activity_before_target.start_of_period, 
+                    next_activity.timestamp
+                ) > 1
+    ) activity_pairs
+"""
 
-        ORDER BY 
-            person_id, 
-            periods.start_of_period ASC
-    ) 
-    
-    WHERE
-        -- exclude first period of dormant
-        ( 
-            (empty(toString(neighbor(person_id, -1))) 
-            OR neighbor(person_id, -1) != person_id
-        ) 
-        AND next_period != {trunc_func}(toDateTime(%(date_from)s) + INTERVAL {interval} - INTERVAL {sub_interval}))
-        OR (
-            (neighbor(person_id, -1) = person_id) 
-            AND neighbor(next_period, -1) < next_period - INTERVAL {interval}
+LIFECYCLE_SQL = f"""
+WITH 
+    %(interval)s AS selected_period,
+
+    -- enumerate all requested periods, so we can zero fill as needed.
```

Comment:
Aside: Please don't use -- in sql comments. This breaks copying queries from metabase, which hurts debugging a lot.

Reply:
We strip comments before exec [here](https://github.com/PostHog/posthog/blob/6fec8bcadd2b3e068833d023ba437b92ca75236a/ee/clickhouse/client.py#L241) so this should be an issue, am I missing something?

Happy to use /* */ if there is still a concern

Reply:
(also, re debugging and metabase, you can use export to csv or similar, which retains new lines. It's not ideal)